bap-builder/PackageMode.go

```diff
@@ -187,29 +182,97 @@ func buildSinglePackage(cmdLine *BuildPackageCmdLineArgs, contextPath string) er
 		ContextPath: contextPath,
 	}
 	packageName := *cmdLine.Name
-	packageJsonDefsList, err := contextManager.GetPackageJsonDefPaths(packageName)
+	packageJsonPathList, err := contextManager.GetPackageJsonDefPaths(packageName)
 	if err != nil {
 		return err
 	}
 
-	for _, packageJsonDef := range packageJsonDefsList {
-		var config bringauto_config.Config
-		err = config.LoadJSONConfig(packageJsonDef)
-		if err != nil {
-			fmt.Fprintf(os.Stderr, "package '%s' JSON config def problem - %s\n", packageName, err)
-			continue
+	logger := bringauto_log.GetLogger()
+
+	var configList []*bringauto_config.Config
+
+	if *cmdLine.BuildDeps {
+		for _, packageJsonPath := range packageJsonPathList {
+			packageJsonPathList = append(packageJsonPathList, getAllDepsJsonPaths(packageJsonPath, contextManager)...)
+		}
+		defsMap := make(ConfigMapType)
+		addConfigsToDefsMap(&defsMap, packageJsonPathList)
+		depList := buildDepList{}
+		configList = depList.TopologicalSort(defsMap)
+	} else {
+		for _, packageJsonPath := range packageJsonPathList {
+			var config bringauto_config.Config
+			err = config.LoadJSONConfig(packageJsonPath)
+			if err != nil {
+				logger.Warn("Couldn't load JSON config from %s path - %s", packageJsonPath, err)
+				continue
+			}
+			configList = append(configList, &config)
 		}
+	}
 
+	for _, config := range configList {
 		buildConfigs := config.GetBuildStructure(*cmdLine.DockerImageName)
+		logger.Info("Build %s", buildConfigs[0].Package.GetFullPackageName())
 		err = buildAndCopyPackage(cmdLine, &buildConfigs)
 		if err != nil {
-			fmt.Fprintf(os.Stderr, "cannot build package '%s' - %s\n", packageName, err)
+			logger.Error("cannot build package '%s' - %s\n", packageName, err)
 			continue
 		}
 	}
 	return nil
 }
 
+func addConfigsToDefsMap(defsMap *ConfigMapType, packageJsonPathList []string) {
+	for _, packageJsonPath := range packageJsonPathList {
+		var config bringauto_config.Config
+		err := config.LoadJSONConfig(packageJsonPath)
+		if err != nil {
+			logger := bringauto_log.GetLogger()
+			logger.Error("Couldn't load JSON config from %s path - %s", packageJsonPath, err)
+			continue
+		}
+		packageName := config.Package.Name
+		_, found := (*defsMap)[packageName]
+		if !found {
+			(*defsMap)[packageName] = []*bringauto_config.Config{}
+		}
+		(*defsMap)[packageName] = append((*defsMap)[packageName], &config)
+	}
+}
+
+func getAllDepsJsonPaths(packageJsonPath string, contextManager ContextManager) []string  {
```

Comment:
Why not to use already existing sorting mechanism from --all veriant?

Reply:
Well, it is just a sorting mechanism. It is used even with  `--build-deps` but firstly we need to get all dependencies recursively for package. This function looks into package config and returns config paths for its dependencies. With `--all` the `ContextManager.GetAllPackagesJsonPath()` method is used, but when building single package, we have only paths for one package.

Reply:
But now as I think about it. Maybe more readable would be add new method to ContextManager, which would return all paths for package and paths for all its dependencies.

Reply:
Yes, it is a good idea to move it to ContextManager.

One more note - did you topological sort dependencies? You load a dependency tree but you do not sort it.
Also, it will be nice to have one topological sort algorithm which is used in both cases (build all and build the package).
Maybe use the same approach as used by --all?

Reply:
Ok, I moved the logic to ContextManager.